### DISCUSSION TITLE
Support Spotify

### DISCUSSION THREAD

**Discussion by birnbeidla** [Tapanila/SharpCaster#308] [Ideas]
I use SharpCaster in a private project and I want to add Spotify support. There is not much out there regarding documentation but there are some libraries (in python) that could help facilitate that. 
My knowledge around chromecast protocol is very limited but I have a brief understanding of C#. Given a little kickstart how one can extend the SharpCaster library with custom channels would help me start my work.

links: 

https://github.com/fondberg/spotcast/blob/master/custom_components/spotcast/spotify_controller.py
https://github.com/home-assistant-libs/pychromecast

**Comment by Tapanila** [comment]
This is really interesting topic. Would you already have the code for getting the Spotify token and etc to start the session or are you mainly interested in controlling already running Spotify session?

I haven't checked how much Spotify is leveraging the chromecast and how much is coming from their server side api

**Reply by Tapanila** [reply]
Just read the content from the link that you sent and I think it seems pretty straightforward. 

I would be fine of adding the Spotify channel to main library. If you want to start a pull request on it I can also help when back from my trip 12th of August.

**Comment by birnbeidla** [comment]
The communication with spotify seems pretty straight forward to me. (Acquire a token from spotify auth service)
My current issue is, that i can't create a proper message with response to get the initial info about the device (see getInfo request in spotify_controller.py)

If you could give me a basic structure for a cutom app channel, i can try to fix up the rest...

**Reply by Tapanila** [reply]
Basically custom channel isn't any different from any other channels. I think MultizoneChannel is the easiest to start from since it's pretty straightforward.

**Reply by Tapanila** [reply]
Also writing a channel especially one that requires taking in events is not the easiest thing to do. Might require quite a bit of knowledge about how chromecast works and especially how I have modeled things.

I don't have any written documentation but I can point on the commits where I added the multizone (it's the latest channel)

**Comment by birnbeidla** [comment]
I tried to implement the "getInfo" message plus response with a little guidance by MultizoneChannel but it didn't work.
I don't know if I did something wrong or if the message is just not working.

I can push my changes but I don't think I'm allowed to create branches or push.

Also I will be off on vacation for the next 2 weeks or so, so this is probably going to be my last post till I return (freshly motivated of course)

**Reply by Tapanila** [reply]
I think timing wise that's perfect. Enjoy your vacation and when you are back I probably will have the Spotify implemented

**Comment by Tapanila** [comment]
Here's initial version:
https://github.com/Tapanila/SharpCaster/tree/spotify

I haven't been able to test this at all so it's probably not working yet

**Reply by Tapanila** [reply]
Now completed the spotify support.
Sample is just connecting the chromecast to spotify account but it doesn't start playing or anything because all of that is then done with Spotify API.

If you have any issues using it please let me know

**Reply by birnbeidla** [reply]
I finally was able to test the spotify support.

Playing works over spotify connect just as expected - as soon as you call addUser, the player appears in the connect api.

Many thanks for your work!!
